### DISCUSSION TITLE
Problems returning from CameraUI on Samsung Galaxy Tab A.

### DISCUSSION THREAD

**Discussion by PhilCase** [airsdk/Adobe-Runtime-Support#1124] [General]
Hi,

One of our customers is having problems with the CameraUI in our SmartCheck app. 

They are using Samsung Galaxy Tab A tablets running Android 11.

They lent us one of the devices to examine the problem and to test an update on. 

I've swapped out the Adobe AIR CameraUI in favour of the distriqt CameraUI which is stable on all other devices I've tested it on bar this one (and also resolves issues with Android 11 file permissions and taking higher resolution photos on some devices).

I believed this improved the usability but the problem persisted that sometimes, on these Samsung Galaxy Tab A devices, the app restarts _sometimes_ after taking a photo (which may be the nth photo in a series of photos).

I can't see anything in the logs that tells me what is going on. I cannot see if it's a problem with the Galaxy Tab A's own Camera App or with Knox blocking access to the Camera or something else.

I can't work out whether it's a problem with The Galaxy Tab A's own camera app.   Michael at distriqt suggested I try a more simple example and test against that.     I've attached the logs from that.

He couldn't see any problems from the distriqt side.   He reported "Nothing stands out there, I can see something weird happening with the air surface view getting removed, but no clear reason as to why."

AIR SDK version: - 33.1.1.575
Device OS: - Android 11 - 32bit
Device version:
Development IDE: - IntelliJ IDEA
Development OS: - Windows 10
Logs: [LOGCAT_DEMO_CAMERAUI.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7104700/LOGCAT_DEMO_CAMERAUI.zip)

If I debug the ActionScript onComplete event - it doesn't get called when the app restarts.   I had the same result here when using the Adobe AIR CameraUI.

I've attached complete logs from the Samsung Galaxy Tab A using the simplified example.   Can you help please?

Thanks.

Phil.

**Comment by marchbold** [comment]
Thats interesting that this is happening with the AIR CameraUI as well. What happens with the AIR implementation? 

What's the exact model on that device, do you know the specs etc? Seems to be a few "Samsung Galaxy Tab A" variants around.

**Comment by PhilCase** [comment]
From the users point of view - the same happens.    From a logcat point of view I'll see if I can generate some logs.

The device is a Samsung Galaxy Tab A 10.1 (2019) LTE - Model: SM-T515.    
Its running Android 11 (AIDA64 says that it's supports 64-bit ARMv8-A (32-bit Mode).   Supported ABIs lists armeabi-v7a, armeabi.)

I'm assuming that it's in 32 bit mode as I wasn't able to install a 64 bit build of the our app on it.

**Comment by ajwfrost** [comment]
Hi

It's certainly running the 32-bit version of AIR.. but can I check, what did you actually do to create the log? It's showing the application start-up and then a 30s delay - perhaps it's looking to connect to a debugger but timing out? - so then at 13:18:45 we see the start of ActionScript processing. There's an error being thrown:
```
09-01 13:18:45.501  7022  7022 I air.uk.co.ikimobile.VideoPlayer: Error while reading build.prop file:Error #3001: File or directory access denied.
```
but I assume that wouldn't cause a problem in your app?

In terms of the visibility of the camera view:
```
09-01 13:18:14.484  7022  7022 I SurfaceView: onWindowVisibilityChanged(0) true com.adobe.air.AndroidCameraView{5f89c0a V.E...... ......I. 0,0-0,0} of ViewRootImpl@af7afac[AppEntry]
09-01 13:18:46.420  7022  7022 I SurfaceView: onWindowVisibilityChanged(8) false com.adobe.air.AndroidCameraView{5f89c0a V.E...... ........ 0,0-8,16} of ViewRootImpl@af7afac[AppEntry]
09-01 13:18:49.675  7022  7022 I SurfaceView: onWindowVisibilityChanged(4) false com.adobe.air.AndroidCameraView{5f89c0a V.E...... ........ 0,0-8,16} of ViewRootImpl@af7afac[AppEntry]
09-01 13:18:49.747  7022  7022 I SurfaceView: onWindowVisibilityChanged(0) false com.adobe.air.AndroidCameraView{5f89c0a V.E...... ........ 0,0-8,16} of ViewRootImpl@af7afac[AppEntry]
```
'Gone' is 8, vs 'Invisible' (4) or 'Visible' (0) so those are all a little weird. I don't suppose we know what happens (from a UI or interaction perspective) at what time in the log?

And is that log just with AIR or when using the Distriqt extension?

When you say that the app restarts, does it automatically start up afresh? Do you see any trace output from your main class constructor? I'm just wondering, since there's also an earlier AIR process happening and this seems to shut down normally, but then we have:
```
09-01 13:18:13.598  3999  4043 I ActivityManager: Start proc 7022:air.uk.co.ikimobile.VideoPlayer/u0a4 for content provider {air.uk.co.ikimobile.VideoPlayer/com.distriqt.extension.cameraui.content.FileProvider}
```
prior to the new process starting up...


FYI there are an increasing number of Samsung devices that are causing weird problems with surfaces and the interactions when we have multiple surfaces present in the app. So we're looking at them as needing a fairly fundamental update to how these are treated....

thanks

**Comment by PhilCase** [comment]
Hi @ajwfrost 

I'm not sure.   It wasn't a release build so might have been initially started from the debugger in IntelliJ IDEA.   I do know it loses connection when the app restarts.

To get the logs I changed to the `33.1.1.575/lib/android/bin` folder after finishing the tests and captured the logs with `adb logcat > file.txt` which is a trick that Andrew at distriqt taught me (I think).

The build.prop file error I get whenever I'm using a "more modern" Android device and appears to be built into the AIR libraries (or possibly flex)   It's not something I'm actively looking for.      I've put it down to trying to read an external file when I don't have permission to and the increased security on the Filesystem.   It doesn't seem to impact the app.   Other Android developers have seen the error also (from searching for it online).

We did get operational problems when we were using the AIR CameraUI.    We were hoping that switching to the distriqt one would help.    We were also getting problems with deleting files from the external CameraRoll caused by enhanced permissions in Android 11 (which Michael fixed - but that was primarily with the video recording functionality).
Operationally - the CameraUI opens, takes it's photo and most of the time returns cleanly.      Sometime it doesn't - which is when the app restarts - the COMPLETE event doesn't get fired at all.  I've only seen it fall over on the Galaxy Tab A though - I'm testing across 4 or 5 phones.
   
Thanks for your input.

**Comment by PhilCase** [comment]
Hi,

I've attached logs from the use of the Adobe CameraUI taking photos in the same application.   This was running on the same Samsung Galaxy Tab A - Model SM-T515.

I'm hoping you'll be able to see that the app crashed twice during the lifetime of the tests.    I didn't think it was going to crash on the second run and I nearly missed it when it restarted.

The app was compiled in release mode for Android 32 using the 33.1.1.575 AIR SDK.   

[TabA.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/7115309/TabA.zip)

Thanks again to you both for your help thus far.

**Reply by ajwfrost** [reply]
I can't see anything in that log that appears to be a crash or an exception, I'm wondering whether it's just that the application is being pushed into the background and therefore not showing up when you've finished taking the photo. Have you ever seen a notification from the Android OS about the application (like, "This app keeps crashing", which is a notification I see if we have 2-3 or more crashes in quick succession..)?

If this app has the AIR/Harman splash screen, do you see that appearing if you go and open it again after this happens?

thanks

**Comment by PhilCase** [comment]
I'm using a multiple view app so when it happens the home view is visible.    

I can build it without the adt.lic file in my AIR library folder to see whether I get the AIR/Harmon splash screen.   (assuming that that controls the licence display).    

Would you expect the app to restart or just crash?   Could it be "just" popping the view stack back to the home view?

Thanks for taking a look.

**Reply by ajwfrost** [reply]
Ah - okay so yes, if the application is still running but just showing a different view, then this is likely to be this problem that some Samsung devices seem to have with z-ordering of views... so I suspect it is screwing up the view stack; if you have any workaround to restore the earlier view back to the top of the display, it may be worth trying that?

thanks

**Reply by marchbold** [reply]
This is one of the things the `blackScreenHelper()` of our Application ANE should help with? Have you tried that. 

I thought it was odd though that the extension isn't getting the returned intent from the camera app, normally if that occurs it indicates something more than a z-ordering issue?

**Reply by PhilCase** [reply]
Would that involve adjusting the navigation to not use the view stack at all - so using replaceView instead of pushView and popView?

I'm not aware of the blackScreenHelper.

**Reply by ajwfrost** [reply]
Hi - I am just working out here, you're talking about pushView/popView etc as Flex capabilities? So I'm not sure how that works but it's all written in ActionScript hence will all be using the single 'stage' view that we have in AIR. The views that I was referring to were Android ones which is what AIR is manipulating...

Just wondering though, are you able to try again using the new build 620 where we made an update to one of the Android views? This has helped with a number of scenarios, although not with everything ...
https://airsdk.harman.com/download/33.1.1.620

thanks

**Reply by PhilCase** [reply]
Yes I'm used to using Flex to manipulate the views.   I haven't looked into doing it in ActionScript without Flex.

I wasn't aware of release 33.1.1.620.    I'll try with that.   thank you.
